server/src/main/java/org/elasticsearch/cluster/coordination/JoinHelper.java

```diff
@@ -189,12 +189,12 @@ public String toString() {
         };
     }
 
-    public void sendJoinRequest(DiscoveryNode destination, Optional<Join> optionalJoin) {
-        sendJoinRequest(destination, optionalJoin, () -> {
-        });
+    boolean isJoinPending() {
+        // cannot use pendingOutgoingJoins.isEmpty() because it's not properly synchronized.
+        return pendingOutgoingJoins.iterator().hasNext();
```

Comment:
The `ConcurrentHashMap` javadocs say:

```
 * Bear in mind that the results of aggregate status methods including
 * {@code size}, {@code isEmpty}, and {@code containsValue} are typically
 * useful only when a map is not undergoing concurrent updates in other threads.
 * Otherwise the results of these methods reflect transient states
 * that may be adequate for monitoring or estimation purposes, but not
 * for program control.
```

Reply:
I believe the weakly consistent only guarantee given by `ConcurrentHashMap` iteration means that following could happen:

1. `pendingOutgoingJoins` has one entry, `e`.
2. A thread `T1` calls `isJoinPending` and gets the iterator (strictly speaking, we halt inside iterator construction before `advance` is called).
3. A thread `T2` adds another entry f and removes e (in that order). So `pendingOutgoingJoins` was never empty.
4. `T1` continues and `hasNext()` can now return `false`.

Mutex protection is done in Coordinator, but is not applied upon receiving requests and responses. It is difficult to see if above scenario can lead to issues, but a simpler solution could be to make the `pendingOutgoingJoins` set synchronized instead (or maybe synchronize on the Coordinator.mutex when manipulating it)?

Reply:
Also from the Javadocs:

```
 * Iterators, Spliterators and Enumerations return elements reflecting the
 * state of the hash table at some point at or since the creation of the
 * iterator/enumeration.
```

The `at some point` in that sentence indicates a snapshot-like semantics which would forbid this situation, and also the consequences are benign as far as I can see, but I'm all for reducing unnecessary mental load. I opened #39900.